File: frontend/src/types/directory.ts
/**
 * Types for directory group mappings, mirroring the backend's `directory` schemas.
 *
 * A mapping names a group as the directory knows it - an LDAP group DN, or a
 * value of an OIDC provider's groups claim - and says what signing in as one of
 * its members makes somebody here: a role in the organization and, optionally,
 * a place in one of its groups.
 */

import type { OrgRole } from "./organization";

export interface DirectoryMapping {
  id: string;
  organization_id: string;
  /** Stored case-folded, which is how the sync compares it. */
  external_group: string;
  role: OrgRole;
  group_id: string | null;
  /** Resolved for display; null when the mapping names no group. */
  group_name: string | null;
  created_at: string;
}

export interface DirectoryMappingList {
  items: DirectoryMapping[];
  total: number;
}

export interface DirectoryMappingCreate {
  external_group: string;
  role: OrgRole;
  group_id: string | null;
}

File: backend/app/schemas/directory.py
"""Schemas for directory group mappings and signing in with a directory account."""

from datetime import datetime
from uuid import UUID

from pydantic import ConfigDict, Field, field_validator

from app.schemas.base import BaseSchema
from app.schemas.organization import InvitableRole


class DirectoryMappingCreate(BaseSchema):
    """Map a directory group to a role, and optionally a group, in this organization."""

    external_group: str = Field(
        min_length=1,
        max_length=1024,
        description=(
            "The group exactly as the identity provider reports it: an LDAP group DN, "
            "or a value of the OIDC groups claim. Compared case-insensitively."
        ),
    )
    role: InvitableRole = Field(
        description="The role a member of the directory group holds here. Never owner."
    )
    group_id: UUID | None = Field(
        default=None, description="A group of this organization the member is also placed in."
    )

    @field_validator("external_group")
    @classmethod
    def case_folded(cls, value: str) -> str:
        """Stored the way it is matched, so the unique constraint means what the match does."""
        return value.casefold()


class DirectoryMappingRead(BaseSchema):
    id: UUID
    organization_id: UUID
    external_group: str
    role: str
    group_id: UUID | None = None
    group_name: str | None = None
    created_at: datetime


class DirectoryMappingList(BaseSchema):
    items: list[DirectoryMappingRead]
    total: int


class DirectoryLogin(BaseSchema):
    """A directory account's own credentials, checked by binding as it.

    Whitespace stripping is switched off for this model: the directory compares a
    password verbatim, so trimming one that ends in a space would refuse its owner
    with the right password - and an account whose password *is* spaces could not
    sign in at all. The username is stripped by its own validator instead.
    """

    model_config = ConfigDict(str_strip_whitespace=False)

    username: str = Field(min_length=1, max_length=256)
    # `min_length=1` refuses the empty password here as well as in the adapter,
    # because an LDAP simple bind with an empty password is an *unauthenticated*
    # bind that most servers answer with success (RFC 4513 s.5.1.2).
    password: str = Field(min_length=1, max_length=1024)
    invitation_handle: str | None = Field(
        default=None,
        max_length=128,
        description=(
            "The handle of an invitation staged before signing in, so a first "
            "directory sign-in on an invite-only deployment is admitted by it"
        ),
    )

    @field_validator("username")
    @classmethod
    def trimmed(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("A username is required")
        return stripped
